builders/base/buildBaseBuilder.sh

```diff
@@ -0,0 +1,147 @@
+    shift 2
+    ;;
+    --buildpack-version)
+    buildpackVersion=$2
+    shift 2
+    ;;
+    --) # end argument parsing
+    shift
+    break
+    ;;
+    -*|--*=) # unsupported flags
+    echo "Error: Unsupported flag $1" >&2
+    exit 1
+    ;;
+    *) # preserve positional arguments
+    PARAMS="$PARAMS $1"
+    shift
+    ;;
+  esac
+done
+
+function replaceRepoWithBuilderMCR() {
+  local imageName="$1"
+  # Retag build image with MCR builder repo
+    IFS=':' read -ra SPLIT_IMAGE_NAME <<< "$imageName"
+    local repo="${SPLIT_IMAGE_NAME[0]}"
+    local tag="${SPLIT_IMAGE_NAME[1]}"
+    local mcrImage="$MCR_BUILDER_IMAGE_REPO:$tag"
+    docker tag "$imageName" "$mcrImage"
+    echo $mcrImage
+}
+
+if [ -z $cliBuilderImage ]; then
+    cliBuilderImage="$destinationFqdn/public/oryx/cli:builder-debian-buster-$builderImageVersion"
+fi
+
+# Create artifact dir & files
+echo "Initializing artifacts file: $ACR_BUILDER_IMAGES_ARTIFACTS_FILE"
+mkdir -p "$ARTIFACTS_DIR/images"
+touch $ACR_BUILDER_IMAGES_ARTIFACTS_FILE
+> $ACR_BUILDER_IMAGES_ARTIFACTS_FILE
+
+# Building stack
+echo "Tagging all images with the tag: $builderImageVersion"
+echo "-------------------------------------------------"
+echo
+echo "Building stack..."
+echo
+baseImage="$destinationFqdn/$destinationRepo:stack-base-$builderImageVersion"
+buildStackImage="$destinationFqdn/$destinationRepo:stack-build-$builderImageVersion"
+runStackImage="$destinationFqdn/$destinationRepo:stack-run-$builderImageVersion"
+
+docker build $SCRIPT_DIR/stack/ \
+    --build-arg CLI_BUILDER_IMAGE="$cliBuilderImage" \
+    -t $baseImage \
+    --target base
+
+docker build $SCRIPT_DIR/stack/ \
+    --build-arg CLI_BUILDER_IMAGE="$cliBuilderImage" \
+    -t $runStackImage \
+    --target run
+
+docker build $SCRIPT_DIR/stack/  \
+    --build-arg CLI_BUILDER_IMAGE="$cliBuilderImage" \
+    -t $buildStackImage \
+    --target build
+
+echo "$baseImage" >> $ACR_BUILDER_IMAGES_ARTIFACTS_FILE
+echo "$runStackImage" >> $ACR_BUILDER_IMAGES_ARTIFACTS_FILE
+echo "$buildStackImage" >> $ACR_BUILDER_IMAGES_ARTIFACTS_FILE
+echo "-------------------------------------------------"
+
+
+# Copy template.buildpack.toml over to buildpack.toml and replace placeholders
+buildpackTomlTemplate="$SCRIPT_DIR/buildpack/template.buildpack.toml"
+targetBuildpackToml="$SCRIPT_DIR/buildpack/buildpack.toml"
+cp "$buildpackTomlTemplate" "$targetBuildpackToml"
+sed -i "s|$ORYX_BUILDPACK_VERSION_PLACEHOLDER|$buildpackVersion|g" "$targetBuildpackToml"
+
+# Packaging buildpack
+buildPackImage="$destinationFqdn/$destinationRepo:buildpack-$builderImageVersion"
+echo
+echo "Packaging buildpack image: $buildPackImage"
+echo
+pack buildpack package $buildPackImage --config $SCRIPT_DIR/packaged-buildpack/package.toml
+echo "$buildPackImage" >> $ACR_BUILDER_IMAGES_ARTIFACTS_FILE
+echo "-------------------------------------------------"
+
+# replace image tags with their MCR equivalent
+mcrBaseImage=$(replaceRepoWithBuilderMCR "$baseImage")
```

Comment:
So this is probably the most confusing part of this whole release pipeline for me. Maybe @cormacpayne can help confirm that this is actually needed...

Basically, when we build this base builder image, we build it using `stack` images that exist in the oryxdevmcr registry. 
_(This is so that in the future, if we would like to add a test suite or something for these images, we can do so. It also allows us to not perform any builds in the `Release` stage of the CI pipeline, as that stage currently just pulls and retags and pushes images for the most part.)_
Currently the builder image was created using `stack` images that exist in the MCR repo, as noted [here](https://github.com/microsoft/Oryx/pull/1934/files#diff-6757db7154ebae953a9876f3831c460cc305817fb5270bbfeff4f4cff8455b26R67). So what these `replaceRepoWith*MCR` methods aim to do is retag the stack images with the MCR repo so they can be used locally with that tag, since _eventually_ we will push those images to MCR anyways in the Release section of the CI pipelines.

My question is whether this is even needed? Is there an issue with just using oryxdevmcr images as the `stack` as long as the final images end up in MCR?

Reply:
Missed this comment when reviewing the PR yesterday -- this is an interesting point because components of the builder depend on specific versions of published images, and it was easy to point to the MCR version of each image when changes were made manually.

As long as the images we're referencing from the `oryxdevmcr` registry are baked along with the builder and there's no additional pulling that needs to be done for these images, then we should be OK. We run into issues when the builder is on MCR and references an `oryxdevmcr` that isn't baked with it and we attempt to pull it down, which causes an authentication issue since `oryxdevmcr` is private (or rather doesn't have anonymous pull enabled).

If it's required that each image referenced amongst the different components must be publicly accessible, then we need to ensure the order of operations is correct here, ideally:
- Build and publish stacks to MCR
- Build and publish buildpack to MCR
  - AFAIK, buildpack doesn't reference specific images, but an `id` that is universal for the stack images
- Update the builder to reference the MCR buildpack and stacks
- Build and publish the builder to MCR

Reply:
Thanks for your comment! Yeah i feel like
`If it's required that each image referenced amongst the different components must be publicly accessible` 
is the crux of the question. I haven't found anything quite yet in CNB docs about whether the various stack images are bundled in with the builder image, or pulled dynamically during runtime. Going to try and do a wee bit of testing to see if I can figure it out

Reply:
Ok testing complete here. Ran:
`pack build pauldtestacr.azurecr.io/testing/container-app:1234 --path ./tests/SampleApps/DotNetCore/NetCore6PreviewMvcApp --builder oryxdevmcr.azurecr.io/public/oryx/builder:Oryx-Validation.20230317.7 --run-image mcr.microsoft.com/oryx/dotnetcore:6.0`

That command uses the image that this CI pipeline produced. Specifically, the stack images that were specified were only tagged with `mcr.microsoft.com` locally in the validation job, and they were never actually pushed to MCR. So if the builder was installing them dynamically during runtime, I should have seen it fail to pull the non-existant images from `mcr.microsoft.com`. Since that didn't happen, I'm concluding that the stack is pre-baked, and I can safely remove the re-tagging logic without ill effects. Does that sound right?

Reply:
That sounds right to me -- my original assumption was that the builder is bundled with all of the images it needs to perform the application build, so if the builder image was built with context to all of these component images, they should be bundled with it (also based on image sizes it made sense the builder included everything else)